### ISSUE TITLE
unable to require jQuery-csv from node

### ISSUE THREAD

**Issue by evanplaice** [evanplaice/jquery-csv#11]
_From [SpencerR...@gmail.com](https://code.google.com/u/118378490492081954495/) on September 27, 2012 07:37:15_

What steps will reproduce the problem? 1. npm install jQuery
2. in node
 - var $ = require('jQuery');
 - var jcsv = require('./jquery.csv.js');

I was getting 

ReferenceError: jQuery is not defined

Whenever I tried this. Some instructions on using this plugin with node would be useful. I figured out that I need to do:

//fake browser window
global.window = require("jsdom").jsdom().createWindow();
global.jQuery = require("jquery");
jcsv = require('jquery.csv.js');

jQuery.csv2Dictionary(stuff); // note that the methods get attached to jQuery, not jcsv

_Original issue: http://code.google.com/p/jquery-csv/issues/detail?id=10_

**Comment by evanplaice** [comment]
_From [evanpla...@gmail.com](https://code.google.com/u/112882755236529658404/) on October 01, 2012 15:53:00_

I honestly have very little experience with Node.js so it will take me a while to get a dev environment setup to to play with this.

I think your problem comes from the way you're calling require. Because jQuery and jquery-csv assume that they're going to be imported into a global namespace by default they take care of the assignment on their own.

If you look at the source, all of the jquery-csv code is contained within an anonymous function closure that is assigned to both the jQuery parameter as well as the $ variable no matter how you call it.

Take a look at http://docs.jquery.com/Plugins/Authoring to see how it works.

Basically, I don't think you need to assign jquery-csv to a variable because it should automatically attach to the jquery namespace.

Try this:
  var $ = require(jquery);
  require('jquery.csv.js');

Both should be callable through the jQuery or the $ variables.

Try it out to see if it works. If not, I'll get a node env setup as soon as I have the time to get a working implementation established.

The key is, if this gets documented as a use case for jquery-csv it needs to work with the client-side documentation/guides as closely as possible.

**Comment by evanplaice** [comment]
_From [evanpla...@gmail.com](https://code.google.com/u/112882755236529658404/) on October 03, 2012 16:13:40_

OK, after further exploration and testing I have a definitive answer.

*Here's how you import jquery-csv:*

    var $ = jQuery = require('jquery');
    require('./jquery.csv.js');

Basically, calling require() without assigning it runs it in the global namespace the same as importing via a \<script> tag does in the browser.

In the browser jQuery is callable via '$' but is actually assigned to the 'jQuery' variable; that's why both assignments are necessary.

If you want to try it out here's the example code I used for testing:
    var fs = require('fs');
    var $ = jQuery = require('jquery');
    require('./jquery.csv.js');

    var file = './example.csv';
    fs.readFile(file, 'UTF-8', function(err, csv) {
      var data = $.csv.toArrays(csv);
      for(var i=0, len=data.length; i<len; i++) {
        console.log(data[i]);
      }
    });

_Note: You'll need to provide your own example.csv file and put it in the same directory as jquery/jquery-csv but it should work._

What it does is read in a csv file, converts it to a 2D array, and outputs the data to the console.

If you want to see how the object transform works, provide headers in the first line of the CSV file and call $.csv.toObjects(csv) instead.

I haven't added any demo's to the project yet but when I do I'll be sure to add this.

**Comment by evanplaice** [comment]
_From [evanpla...@gmail.com](https://code.google.com/u/112882755236529658404/) on October 03, 2012 16:14:24_

**Status:** Done

**Comment by evanplaice** [comment]
_From [SpencerR...@gmail.com](https://code.google.com/u/118378490492081954495/) on October 04, 2012 05:20:52_

Thanks for fixing this. I moved on to using the node-csv-parser though, since its API fits my use case better.

Link here: https://github.com/wdavidw/node-csv-parser

**Comment by evanplaice** [comment]
_From [evanpla...@gmail.com](https://code.google.com/u/112882755236529658404/) on October 04, 2012 15:17:41_

No problem...

Now that I have a Node.js environment up and running I'm going to see if I can add some solid callback support to make it work in a manner familiar to most Node programmers.

Node isn't really something I have considered up until you submitted this issue. I think it'll be interesting to add a demo showing how to implement a 100% JS CSV SaaS (ie client/server) module.

Thank you for your feedback.

**Comment by evanplaice** [comment]
_From [evanpla...@gmail.com](https://code.google.com/u/112882755236529658404/) on October 10, 2012 21:08:00_

**Owner:** evanpla...@gmail.com
